1-Draft/RFC00XX-Native-Command-Error-Handling.md

```diff
@@ -0,0 +1,229 @@
+
+`Set -eo pipefail` is a combination of `Set -u` and `Set -o pipefail`:
+
+- `Set -u`, a reference to any variable not previously defined results in an error. Similar to
+  Set-StrictMode in PowerShell.
+- `Set -o pipefail`, by default, pipeline success is determined by the last command executed. `Set
+  -o pipefail` returns an error if any command in the pipeline fails.
+
+ As a result of not including the semantics of `Set -eo pipefail`, PowerShell script logic for
+ handling native command exit codes would need to use either `try`..`catch` or existing exit status
+ handling language constructs, according to the setting of this preference variable.
+
+ One way of overriding `$ErrorActionPreference` for a single native command and handling
+ it's exit status explicitly would be to put this logic into a script block and call it
+ with the invocation operator (`&`).
+
+## Alternative Approaches and Considerations
+
+We could extend this existing behavior to give equivalent functionality to bash `Set -eo pipefail`.
+This includes:
+
+- `Set -u`, a reference to any variable not previously defined results in an error. Similar to
+  `Set-StrictMode` in PowerShell.
+- `Set -o pipefail`, by default, pipeline success is determined by the last command executed. `Set
+  -o pipefail` returns an error if any command in the pipeline fails.
+
+Today there exists a workaround for handling native command exit codes using a `try`..`catch` block.
+This alternative may be considered in the future.
+
+### Add "strict" native command option
+
+The `$PSStrictNativeCommand` preference should treat creation of an `ErrorRecord` for native
+commands in the same way as this is treated elsewhere. Described here as a Boolean, could be
+considered as an enum to allow for future expansion. Possible values are:
+
+- `$false`: (the default) ignore non-zero exit codes. This is the same as existing PowerShell
+  treatment of this case.
+- `$true`: Populate the error stream of the native command with an `ErrorRecord` associated with an
+  `ExitException` exception.
+
+### Modifying existing semantics to consider exit code and exit status
+
+The error will throw an exception, potentially terminating the session, in the same situation as
+other non-terminating errors will do this, i.e. where `$ErrorActionPreference` is set to `"stop"`.
+This would not be the desired behavior on commands where the script already handles a non-zero exit
+code, which would require the addition of extra boilerplate to use multiple native commands in
+combination. There are a number of ways that this could be made more flexible by integrating exit
+code and exit status handling into the language syntax.
+
+#### Convert non-terminating errors to terminating where the command output is used
+
+This approach implements semantics equivalent to bash `set -eo pipefail` in the runtime layer.
+
+The `$PSStrictPipeLine` preference variable would govern promotion of a non-terminating error to a
```

Comment:
Seems like we could just have a different value to the enum for `$PSNativeCommandErrorAction` instead of a new preference variable?